src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                {
+                    return false;
+                }
+
+                BindingDiagnosticBag applicableDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
+                if (!Conversions.IsApplicableInterpolatedStringBuilderType(unconvertedInterpolatedString, interpolatedStringBuilderType, applicableDiagnostics, out var builderArguments))
+                {
+                    applicableDiagnostics.Free();
+                    return false;
+                }
+
+                diagnostics.AddRangeAndFree(applicableDiagnostics);
+
+                // Prior to C# 10, all types in an interpolated string expression needed to be convertible to `object`. After 10, some types
+                // (such as Span<T>) that are not convertible to `object` are permissible as interpolated string components, provided there
+                // is an applicable TryFormatInterpolationHole that accepts them. To preserve langversion, we therefore make sure all components
+                // are convertible to object if the current langversion is lower than the interpolation feature
+                TypeSymbol? objectType = null;
+                BindingDiagnosticBag? conversionDiagnostics = null;
+                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+                if (needToCheckConversionToObject)
+                {
+                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                    conversionDiagnostics = BindingDiagnosticBag.GetInstance();
+                }
+
+                // Swap out the first argument of the format calls, which is the (potentially converted) part from the original string with
+                // a placeholder, and put the part into the interpolated string's list of parts.
+                Debug.Assert(builderArguments.Length == unconvertedInterpolatedString.Parts.Length);
+                var interpolatedDataBuilder = ArrayBuilder<MethodArgumentInfo>.GetInstance(builderArguments.Length);
+                var argumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(3);
+                var partsBuilder = ArrayBuilder<BoundExpression>.GetInstance(builderArguments.Length);
+                int baseStringLength = 0;
+                int numFormatHoles = 0;
+
+                for (int i = 0; i < builderArguments.Length; i++)
+                {
+                    var currentFormatCall = builderArguments[i];
+                    Debug.Assert(currentFormatCall.Arguments.Length > 0);
+                    Debug.Assert(argumentsBuilder.Count == 0);
+
+                    var newPart = currentFormatCall.Arguments[0];
+                    argumentsBuilder.Add(new BoundInterpolatedStringElementPlaceholder(newPart.Syntax, i, newPart.Type));
```

Comment:
>argumentsBuilder.Add(new BoundInterpolatedStringElementPlaceholder(newPart.Syntax, i, newPart.Type)); [](start = 20, length = 101)

I am not sure why we are doing this. It also looks like an assumption is made that the part and coerced argument are the same thing, whereas I don't think this is necessary the case. I think we either should bind againt a placeholder from the very beginning and then use standard substitution during lowering as we do with other placeholders, or we should leave the argument alone and just lower it normally.

Reply:
>  I think we either should bind againt a placeholder from the very beginning

This won't work with target-typed expressions.

---
In reply to: [610901304](https://github.com/dotnet/roslyn/pull/52422#discussion_r610901304) [](ancestors = 610901304)